packages/devextreme/testing/tests/DevExpress.ui.widgets/tabs.tests.js

```diff
@@ -422,6 +423,200 @@ QUnit.module('General', () => {
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 before promise resolves');
+            assert.deepEqual(this.tabInstance.option('selectedItem'), { text: '0' }, 'selectedItem should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItems'), [{ text: '0' }], 'selectedItems should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItemKeys'), [{ text: '0' }], 'selectedItemKeys should remain the first key');
+
+            this.clock.tick(this.CANCEL_DELAY);
+
+            onSelectionChangingSpy.getCall(0).args[0].cancel.catch((error) => {
+                assert.strictEqual(error, 'Cancellation error', 'Promise should be rejected with "Cancellation error"');
+                assert.strictEqual(onSelectionChangingSpy.callCount, 1, 'onSelectionChanging should still be called once');
+                assert.strictEqual(this.onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should be called after promise rejection');
+                assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should change after cancellation promise rejection');
+                assert.deepEqual(this.tabInstance.option('selectedItem'), { text: '0' }, 'selectedItem should remain the first item');
+                assert.deepEqual(this.tabInstance.option('selectedItems'), [{ text: '0' }], 'selectedItems should remain the first item');
+                assert.deepEqual(this.tabInstance.option('selectedItemKeys'), [{ text: '0' }], 'selectedItemKeys should remain the first key');
+            });
+        });
+
+        QUnit.test('Selection should be cancelled asynchronously when e.cancel is resolved', function(assert) {
+            const onSelectionChangingSpy = sinon.spy(function(e) {
+                e.cancel = new Promise((resolve) => {
+                    setTimeout(() => {
+                        resolve(true);
+                    }, this.CANCEL_DELAY);
+                });
+            });
+
+            this.tabInstance.option('onSelectionChanging', onSelectionChangingSpy);
+            const $item = this.$tabs.find(`.${TABS_ITEM_CLASS}`).eq(1);
+
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'Initial selectedIndex should be 0');
+            assert.strictEqual(onSelectionChangingSpy.callCount, 0, 'onSelectionChanging should not be called initially');
+
+            $item.trigger('dxclick');
+
+            this.clock.tick(this.CANCEL_DELAY);
+
+            assert.strictEqual(onSelectionChangingSpy.callCount, 1, 'onSelectionChanging should be called once after click');
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 before promise resolves');
+            assert.ok(isPromise(onSelectionChangingSpy.getCall(0).args[0].cancel), 'e.cancel should be a Promise');
```

Comment:
[excess code]
This assert checks nothing: u set cancel to promise manually => it cannot be not a promise
```suggestion
```